docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
```

Comment:
It does seem to me that `r_<role_name>` would be best reserved for inputs and outputs from a role. (You could *even* consider a different prefix for outputs, say `o_<role_name>`. Or not.)

I think there are two use cases for "local" variables (though there's no such thing as a local variable). One is the `l_result` pattern where a task sets the variable and then immediately other tasks in the same task file use it, and it's really easy to see that the variable won't accidentally get a value from somewhere unexpected. The other is the pattern where a variable is set internally and meant to be used in other places in the role, for example in a template, but it's not as simple to determine (visually) whether and where it was set, and there's reasonable risk that you might include tasks from another role or even the same role that could use the same name and overwrite what you thought you were getting.

So here's an idea. As in the python tradition, prefix local variables with `_`. If it's a variable that has some risk of conflicting with another "local" variable, further prefix it with the role name, so `_<role_name>_internal_value`. We could consider prefixing the "super local" variables with another `_` so e.g. `__result`.